core/blockchain_pruner.go

```diff
@@ -0,0 +1,227 @@
+	err := bp.batchWriter.Delete(key)
+	if err != nil {
+		return err
+	}
+
+	if bp.batchWriter.ValueSize() >= ethdb.IdealBatchSize {
+		err = bp.batchWriter.Write()
+		if err != nil {
+			return err
+		}
+
+		bp.batchWriter.Reset()
+	}
+	return nil
+}
+
+func (bp *blockchainPruner) resetBatchWriter() error {
+	err := bp.batchWriter.Write()
+	if err != nil {
+		return err
+	}
+
+	bp.batchWriter = nil
+	return nil
+}
+
+func (bp *blockchainPruner) runWhenNotPruning(cb func() error) error {
+	// check is start
+	if atomic.CompareAndSwapInt32(&bp.isPruneBeaconChainRunning, 0, 1) {
+		defer atomic.StoreInt32(&bp.isPruneBeaconChainRunning, 0)
+
+		return cb()
+	}
+	return nil
+}
+
+func (bp *blockchainPruner) Start(maxBlockNum uint64, maxEpoch *big.Int) error {
+	return bp.runWhenNotPruning(func() error {
+		// init
+		bp.compactRange = make([][2][]byte, 0)
+		bp.startTime = time.Now()
+		bp.deletedBlockCount = 0
+		bp.deletedValidatorSnapshot = 0
+		bp.skipValidatorSnapshot = 0
+		bp.batchWriter = bp.db.NewBatch()
+
+		// prune data
+		bp.addToCompactRange(bp.pruneBeaconChainBlock(maxBlockNum))
+		bp.addToCompactRange(bp.pruneBeaconChainValidatorSnapshot(maxEpoch))
+
+		//  batch write data and reset
+		err := bp.resetBatchWriter()
+		if err != nil {
+			return err
+		}
+
+		prunerMaxBlock.Set(float64(maxBlockNum))
+		deletedValidatorSnapshot.Add(float64(bp.deletedValidatorSnapshot))
+		skipValidatorSnapshot.Add(float64(bp.skipValidatorSnapshot))
+		deletedBlockCount.Add(float64(bp.deletedBlockCount))
+		deletedBlockCountUsedTime.Add(float64(time.Now().Sub(bp.startTime).Milliseconds()))
+
+		utils.Logger().Info().
+			Uint64("maxBlockNum", maxBlockNum).
+			Uint64("maxEpoch", maxEpoch.Uint64()).
+			Int("deletedBlockCount", bp.deletedBlockCount).
+			Int("deletedValidatorSnapshot", bp.deletedValidatorSnapshot).
+			Int("skipValidatorSnapshot", bp.skipValidatorSnapshot).
+			Dur("cost", time.Now().Sub(bp.startTime)).
+			Msg("pruneBeaconChain delete block success")
+
+		// probability of 1 in 1000 blocks to Compact, It consumes a lot of IO
+		if rand.Intn(compactProportion) == 1 {
+			startTime := time.Now()
+			for _, compactStartEnd := range bp.compactRange {
+				err := bp.db.Compact(compactStartEnd[0], compactStartEnd[1])
+				if err != nil {
+					return err
+				}
+			}
+			compactBlockCountUsedTime.Add(float64(time.Now().Sub(startTime).Milliseconds()))
+			utils.Logger().Info().
+				Uint64("maxBlockNum", maxBlockNum).
+				Uint64("maxEpoch", maxEpoch.Uint64()).
+				Dur("cost", time.Now().Sub(startTime)).
+				Msg("pruneBeaconChain compact db success")
+		}
+
+		return nil
+	})
+}
+
+func (bp *blockchainPruner) pruneBeaconChainBlock(maxBlockNum uint64) (minKey []byte, maxKey []byte) {
+	return rawdb.IteratorBlocks(bp.db, func(blockNum uint64, hash common.Hash) bool {
+		if blockNum >= maxBlockNum {
+			return false
+		}
+
+		// skip genesis block
+		if blockNum == 0 {
+			return true
+		}
+
+		blockInfo := rawdb.ReadBlock(bp.db, hash, blockNum)
```

Comment:
Looks like if a certain block is pruned in last iteration, block Info can be nil. Could this possibly happen? If we reach a block that is pruned previously, does the prune iteration continue or break?